ark/type/src/parser/objectLiteral.ts

```diff
@@ -1,85 +1,109 @@
-import { PredicateNode } from "../nodes/predicate/predicate.js"
-import { DomainNode } from "../nodes/primitive/domain.js"
-import type { NamedPropsInput } from "../nodes/prop/props.js"
-import { PropsNode } from "../nodes/prop/props.js"
-import { TypeNode } from "../nodes/type.js"
 import type { ParseContext } from "../scope.js"
 import type { inferDefinition, validateDefinition } from "./definition.js"
 import type { validateString } from "./semantic/validate.js"
-import { Scanner } from "./string/shift/scanner.js"
+import {
+	IndexedKey,
+	OptionalValue,
+	parseKeyValueEntry,
+	parsedEntry,
+	validateObjectValue
+} from "./shared.js"
+import type { parseEntry } from "./shared.js"
+import {
+	type Dict,
+	type error,
+	type evaluate,
+	type mutable
+} from "@arktype/util"
 
 export const parseObjectLiteral = (def: Dict, ctx: ParseContext) => {
 	const named: mutable<NamedPropsInput> = {}
+
 	for (const definitionKey in def) {
-		let keyName = definitionKey
-		let optional = false
-		if (definitionKey[definitionKey.length - 1] === "?") {
-			if (definitionKey[definitionKey.length - 2] === Scanner.escapeToken) {
-				keyName = `${definitionKey.slice(0, -2)}?`
-			} else {
-				keyName = definitionKey.slice(0, -1)
-				optional = true
-			}
-		}
-		ctx.path.push(keyName)
-		named[keyName] = {
-			prerequisite: false,
-			optional,
-			value: ctx.scope.parse(def[definitionKey], ctx)
-		}
-		ctx.path.pop()
+		addKeyValuePairToNamed(def, definitionKey, named, ctx)
+	}
+	const symbols = Object.getOwnPropertySymbols(def)
+	for (const symbol of symbols) {
+		addKeyValuePairToNamed(def, symbol, named, ctx)
 	}
 	// TODO: meta
 	return node({ basis: "object", props: named }, ctx)
 }
 
+const addKeyValuePairToNamed = (
+	def: Dict,
+	definitionKey: string | symbol,
+	named: mutable<NamedPropsInput>,
+	ctx: ParseContext
+) => {
+	const { innerKey, innerValue, optional } = parseKeyValueEntry(
+		definitionKey,
+		def[definitionKey]
+	)
+	ctx.path.push(innerKey as string)
+	const definition = ctx.scope.parse(innerValue, ctx)
+
+	named[innerKey as string] = {
+		prerequisite: false,
+		value: definition,
+		optional
+	}
+	ctx.path.pop()
+}
+
 export type inferObjectLiteral<def extends object, $, args> = evaluate<
 	{
 		// since def is a const parameter, we remove the readonly modifier here
 		// support for builtin readonly tracked here:
 		// https://github.com/arktypeio/arktype/issues/808
 		-readonly [k in keyof def as nonOptionalKeyFrom<
 			k,
+			def[k],
 			$,
 			args
 		>]: inferDefinition<def[k], $, args>
 	} & {
-		-readonly [k in keyof def as optionalKeyFrom<k>]?: inferDefinition<
-			def[k],
+		-readonly [k in keyof def as optionalKeyFrom<k, def[k]>]?: inferDefinition<
+			def[k] extends OptionalValue<infer inner> ? inner : def[k],
 			$,
 			args
 		>
 	}
 >
 
-type nonOptionalKeyFrom<k, $, args> = parseKey<k> extends {
+type nonOptionalKeyFrom<k extends PropertyKey, valueDef, $, args> = parseEntry<
+	k,
+	valueDef
+> extends parsedEntry<{
```

Comment:
Now that we have a ParsedEntry type, it will be easier to just `infer result extends ParsedEntry`, then access the props you need as `result["innerKey"]` etc. so you can avoid duplicating it